tests/acme_endorsement.py

```diff
@@ -0,0 +1,341 @@
+                        context.load_verify_locations(cadata=crt)
+
+                    s = socket()
+                    s.settimeout(1)
+                    c = context.wrap_socket(s, server_hostname=network_name)
+                    c.connect((iface.host, iface.port))
+                    cert_der = c.getpeercert(binary_form=True)
+                    cert = load_der_x509_certificate(cert_der, default_backend())
+                    if cert.subject.rfc4514_string() != "CN=" + network_name:
+                        raise Exception("Common name mismatch")
+                    cert_public_key = cert.public_key().public_bytes(
+                        encoding=Encoding.PEM, format=PublicFormat.SubjectPublicKeyInfo
+                    )
+                    if network_public_key != cert_public_key:
+                        raise Exception("Public key mismatch")
+                    num_ok += 1
+                except Exception as ex:
+                    LOG.trace(f"Likely expected exception: {ex}")
+
+            if num_ok != len(args.nodes):
+                time.sleep(1)
+
+        LOG.info(
+            f"Success: all nodes had correct certificates installed after {int(time.time() - start_time)} seconds"
+        )
+
+
+def get_binary(url, filename):
+    if not os.path.isfile(filename):
+        urllib.request.urlretrieve(
+            url,
+            filename,
+        )
+        os.chmod(filename, 0o744)
+
+
+def start_mock_dns(filename, listen_address, mgmt_address, out, err, env=None):
+    p = subprocess.Popen(
+        [
+            filename,
+            "-http01",
+            "",
+            "-https01",
+            "",
+            "-dns01",
+            listen_address,
+            "-tlsalpn01",
+            "",
+            "-management",
+            mgmt_address,
+        ],
+        stdout=out,
+        stderr=err,
+        close_fds=True,
+        env=env,
+    )
+    time.sleep(1)
+    return p
+
+
+def register_endorsed_hosts(args, network_name, dns_mgmt_address):
+    endorsed_hosts = [
+        node.rpc_interfaces["acme_endorsed_interface"].host for node in args.nodes
+    ]
+    data = str(
+        json.dumps(
+            {
+                "host": network_name,
+                "addresses": endorsed_hosts,
+            }
+        )
+    ).encode("utf-8")
+    urllib.request.urlopen("http://" + dns_mgmt_address + "/add-a", data=data)
+
+    # Disable the default IPv6 entry
+    data = str(json.dumps({"ip": ""})).encode("utf-8")
+    urllib.request.urlopen(
+        "http://" + dns_mgmt_address + "/set-default-ipv6", data=data
+    )
+
+
+def start_pebble(filename, config_filename, dns_address, out, err, env):
+    p = subprocess.Popen(
+        [
+            filename,
+            "--config",
+            config_filename,
+            "-dnsserver",
+            dns_address,
+        ],
+        stdout=out,
+        stderr=err,
+        close_fds=True,
+        env=env,
+    )
+    time.sleep(2)
```

Comment:
That's potentially brittle, it'd be better to wait for a condition, like a line printed to stdout, or the OS knowing something is listening on the socket for example.